homeassistant/components/zestimate/sensor.py

```diff
@@ -113,12 +113,16 @@ def update(self):
             return
         data = data_dict['response'][NAME]
         details = {}
-        details[ATTR_AMOUNT] = data['amount']['#text']
-        details[ATTR_CURRENCY] = data['amount']['@currency']
-        details[ATTR_LAST_UPDATED] = data['last-updated']
-        details[ATTR_CHANGE] = int(data['valueChange']['#text'])
-        details[ATTR_VAL_HI] = int(data['valuationRange']['high']['#text'])
-        details[ATTR_VAL_LOW] = int(data['valuationRange']['low']['#text'])
+        if 'amount' in data and data['amount'] is not None:
+            details[ATTR_AMOUNT] = data['amount']['#text']
+            details[ATTR_CURRENCY] = data['amount']['@currency']
+        if 'last-updated' in data and data['last-updated'] is not None:
+            details[ATTR_LAST_UPDATED] = data['last-updated']
+        if 'valueChange' in data and data['valueChange'] is not None:
+            details[ATTR_CHANGE] = int(data['valueChange']['#text'])
+        if 'valuationRange' in data and data['valuationRange'] is not None:
+            details[ATTR_VAL_HI] = int(data['valuationRange']['high']['#text'])
+            details[ATTR_VAL_LOW] = int(data['valuationRange']['low']['#text'])
```

Comment:
```
for attr, key, value in ((ATTR_XY, 'amount', '#text'),):
    if data.get(key) is None:
        continue
    details[attr] = data[key][value]
```

Reply:
thx @pvizeli  this suggestion is probably a bit cleaner but I'm not sure how I would write it without it being a lot more code.  The two challenges are that the homeassistant attribute names don't match the XML attributes names coming in so they have to be translated (otherwise this would become a breaking change).  Also, it's pulling attributes at different levels from the incoming XML structure so I think it would have to be multiple for loops to get all the data.  

something like this?....

```
        for attr, key, value in ((data, 'amount', '#text'),):
            if data.get(key) is None:
                continue
            details[ATTR_AMOUNT] = data[key][value]
            details[ATTR_CURRENCY] = data[key]['@currency']
        for attr, key, value in ((data, 'last-updated', '#text'),):
            if data.get(key) is None:
                continue
            details[ATTR_LAST_UPDATED] = data[key]
        for attr, key, value in ((data, 'valueChange', '#text'),):
            if data.get(key) is None:
                continue
            details[ATTR_CHANGE] = data[key][value]
        for attr, key, value in ((data, 'valuationRange', '#text'),):
            if data.get(key) is None:
                continue
            ...
            ...
            ...
```
The    incoming XML looks like this...
```
<response>
...
...
    <address>
        <street>123 Misery Dr</street>
        <zipcode>44240</zipcode>
        <city>Bent</city>
        <state>VA</state>
        <latitude>41.160999</latitude>
        <longitude>-81.384689</longitude>
    </address>
    <zestimate>
        <amount currency="USD">229093</amount>
        <last-updated>04/22/2019</last-updated>
        <oneWeekChange deprecated="true"/>
        <valueChange/>
        <valuationRange>
            <low currency="USD">217638</low>
            <high currency="USD">240548</high>
        </valuationRange>
        <percentile>90</percentile>
    </zestimate>
...
...
</response>
```